### PR TITLE
Add kernels for Allen-Cahn equation in multi-phase KKS system

### LINKED ISSUE #6705: Add kernels for multi-order parameter KKS phase-field model
### Description of the enhancement or error report
Add `KKSMultiACBulkBase`, `KKSMultiACBulkF`, and `KKSMultiACBulkC` kernels for multi-order parameter KKS phase-field models

### Rationale for the enhancement or information for reproducing the error
Similar to the single-order parameter KKS phase-field model, the `df/deta_i` terms are rewritten using chain rule expressions, so these terms are not obtained from derivative materials.

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)
None

### PR DESCRIPTION
**Pull request by laagesen**
Closes #6705 

These kernels are used for both the Allen-Cahn equations (where the order parameter you take the derivative w.r.t. is the nonlinear variable) and the equation for the Lagrange multiplier (where the Lagrange multiplier is the nonlinear variable). Therefore, in the case where they are used in the Lagrange multiplier equation, you need to specify what order parameter `eta_i` to take the derivative w.rt. To prevent confusion for the user, I made `eta_i` required. You could make it an optional parameter then if it's not specified, default to using the kernel's nonlinear variable as `eta_i`, but that would create the risk that the user would fail to specify `eta_i` when using these kernels for the Lagrange multiplier equation, where it's needed. I'm open to suggestion if there's a better way to guard against that happening, though...

### CODE REVIEW

File: modules/phase_field/src/kernels/KKSMultiACBulkC.C [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+Real
+KKSMultiACBulkC::computeDFDOP(PFFunctionType type)
+{
+  Real sum = 0.0;
+
+  switch (type)
+  {
+    case Residual:
+    {
```
Comment (dschwen) [inline modules/phase_field/src/kernels/KKSMultiACBulkC.C]: we only use curly braces around case blocks if we need to create a new _scope_.

File: modules/phase_field/src/kernels/KKSMultiACBulkC.C [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+      for (unsigned int n = 0; n < _num_j; ++n)
+        sum += (*_prop_dhjdetai[n])[_qp] * (*_cjs[n])[_qp];
+
+      return - _prop_dF1dc1[_qp] * sum;
```
Comment (dschwen) [inline modules/phase_field/src/kernels/KKSMultiACBulkC.C]: we don't leave a space after unary minus.

File: modules/phase_field/src/kernels/KKSMultiACBulkC.C [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+  // member function
+  Real res = ACBulk<Real>::computeQpOffDiagJacobian(jvar);
+
+  Real sum = 0.0;
+  // Then add dependence of KKSACBulkC on other variables
+  // Treat cj variables specially, as they appear in the residual
+
+  for (unsigned int i = 0; i < _num_j; ++i)
+  {
+    if (jvar == _cjs_var[0])
```
Comment (dschwen) [inline modules/phase_field/src/kernels/KKSMultiACBulkC.C]: This condition and its block do not depend on `i`. Move before the loop.

File: modules/phase_field/tests/KKS_system/tests [hunk 4]
```diff
@@ -27,5 +27,6 @@
     type = 'Exodiff'
     input = 'kks_multiphase.i'
     exodiff = 'kks_multiphase_out.e'
+    max_parallel = 1                # b/c of -pc_type lu
```
Comment (dschwen) [inline modules/phase_field/tests/KKS_system/tests]: Nope, pc_type is asm!

**Comment by dschwen** [comment]
I'll merge this once you confirm that it passes the Jacobian tester.

**Comment by laagesen** [comment]
OK, previously I had adjusted the on-diagonal Jacobian calculation for the case where the kernel is used in the Lagrange multiplier equation rather than an Allen-Cahn equation - but I neglected to make the corresponding change to the off-diagonal Jacobian calculation for that case. So a few changes coming...

**Comment by laagesen** [comment]
Also, I'm adding a check to make sure the user specifies the `args` needed for the switching function materials, which now depend on all the order parameters. (this is an addition to the existing check that the phase compositions are supplied as args for the free energies)

**Comment by dschwen** [comment]
Could you squash this down a bit?

**Pull request merged**